tests/bluetooth/bsim_bt/bsim_test_l2cap/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,602 @@
+}
+
+static void test_peripheral_main(void)
+{
+	device_sync_init(PERIPHERAL_ID);
+	LOG_DBG("*L2CAP ECRED Peripheral started*");
+	init_workqs();
+	int err = bt_enable(NULL);
+
+	if (err) {
+		FAIL("Can't enable Bluetooth (err %d)", err);
+		return;
+	}
+
+	LOG_DBG("Peripheral Bluetooth initialized.");
+	LOG_DBG("Connectable advertising...");
+	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
+
+	if (err) {
+		FAIL("Advertising failed to start (err %d)", err);
+		return;
+	}
+
+	LOG_DBG("Advertising started.");
+	LOG_DBG("Peripheral waiting for connection...");
+	WAIT_FOR_FLAG_SET(is_connected);
+	LOG_DBG("Peripheral Connected.");
+	register_l2cap_server();
+	connect_num_channels(L2CAP_CHANNELS);
+	k_sem_take(&all_chan_conn_sem, K_FOREVER);
+
+	/* Disconnect and reconnect channels *****************************************************/
+	LOG_DBG("############# Disconnect and reconnect channels");
+	disconnect_all_channels();
+	k_sem_take(&all_chan_disconn_sem, K_FOREVER);
+
+	connect_num_channels(L2CAP_CHANNELS);
+	k_sem_take(&all_chan_conn_sem, K_FOREVER);
+
+	LOG_DBG("Send sync after reconnection");
+	device_sync_send();
+
+	/* Send bytes on both channels and expect ch 1 to receive all of them before ch 0 *********/
+	LOG_DBG("############# Send bytes on both channels concurrently");
+	send_sdu_concurrently();
+
+	/* Change mtu size on all connected channels *********************************************/
+	LOG_DBG("############# Change MTU of the channels");
+	err = change_mtu_on_channels(L2CAP_CHANNELS, CONFIG_BT_L2CAP_TX_MTU + 10);
+
+	if (err) {
+		FAIL("MTU change failed (err %d)\n", err);
+	}
+
+	/* Read from both devices (Central and Peripheral) at the same time **********************/
+	LOG_DBG("############# Read from both devices (Central and Peripheral) at the same time");
+	LOG_DBG("Wait for sync before sending the msg");
+	device_sync_wait();
+	LOG_DBG("Received sync");
+	send_sdu(0, 1, 10);
+
+	k_sem_take(&sent_sem, K_FOREVER);
+	disconnect_all_channels();
+	WAIT_FOR_FLAG_UNSET(is_connected);
+	PASS("L2CAP ECRED Peripheral tests Passed");
+	bs_trace_silent_exit(0);
+}
+
+static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
+			 struct net_buf_simple *ad)
+{
+	struct bt_le_conn_param *param;
+	int err = bt_le_scan_stop();
+
+	if (err) {
+		FAIL("Stop LE scan failed (err %d)", err);
+		return;
+	}
+
+	param = BT_LE_CONN_PARAM_DEFAULT;
+	err = bt_conn_le_create(addr, BT_CONN_LE_CREATE_CONN, param, &default_conn);
+	if (err) {
+		FAIL("Create conn failed (err %d)", err);
+		return;
+	}
+}
+
+static void test_central_main(void)
+{
+	struct bt_le_scan_param scan_param = {
+		.type = BT_LE_SCAN_TYPE_ACTIVE,
+		.options = BT_LE_SCAN_OPT_NONE,
+		.interval = BT_GAP_SCAN_FAST_INTERVAL,
+		.window = BT_GAP_SCAN_FAST_WINDOW,
+	};
+
+	device_sync_init(CENTRAL_ID);
+
+	LOG_DBG("*L2CAP ECRED Central started*");
+
+	int err = bt_enable(NULL);
+
+	if (err) {
```

Comment:
Ok, I can see this is a recurring pattern. I'd feel perhaps less strongly about tying the assignment to the declaration if there was a single use of the variable, but since you're using it for multiple function calls it doesn't quite feel right to special-case the first one as part of the declaration.